### DISCUSSION TITLE
Welcome to LuisaRender Discussions!

### DISCUSSION THREAD

**Discussion by Mike-Leo-Smith** [LuisaGroup/LuisaRender#4] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by odil24** [comment]
Hi. My name is Odilkhan Yakubov aka odil24
This is great that you are created this dis thread

**Comment by odil24** [comment]
are you planned to make an exporter for Blender also?

**Reply by Mike-Leo-Smith** [reply]
Yes, exporters and other more “production-oriented” stuff are in the long-term plan. But at present, we mainly focus on the development of the renderer itself, since there are still many features to complete.

**Reply by odil24** [reply]
> Yes, exporters and other more “production-oriented” stuff are in the long-term plan. But at present, we mainly focus on the development of the renderer itself, since there are still many features to complete.

Is there have any build for testing render engine?

**Reply by Mike-Leo-Smith** [reply]
We don't make binary releases currently, but I can invite you into the "Insiders" team of LuisaGroup so that you can have access to the source including LuisaCompute.

**Reply by odil24** [reply]
> We don't make binary releases currently, but I can invite you into the "Insiders" team of LuisaGroup so that you can have access to the source including LuisaCompute.

If so can you give instructions on how to compile from the source code step by step?

**Comment by orarugogerincufelpattano** [comment]
Hi! 
The last comment was nearly two years ago so I'm curious how the development of the renderer and a possible Blender Exporter is proceeding.

**Reply by Mike-Leo-Smith** [reply]
The renderer has gone through quite some enhancements and internal refactoring. For example, we changed how surface closures are expanded to `switch-case` statements; more integrators, like SPPM, Gradient-Domain PT, etc are added; and, many changes from the underlying LuisaCompute framework are consistently being merged, like Ray Query for efficient alpha tests.

For the exporter, I still don't have a specific plan. My friend has made some good progress in his renderer's blender [exporter](https://github.com/shiinamiyuki/akari_render/blob/next/crates/akari_scenegraph/python/exporter.py). I might try to see if it's possible to reuse his work.

**Reply by orarugogerincufelpattano** [reply]
Thanks for the Infos! 
Unfortunately I can't try the rendered as I have no admin access on my work pc.
Do you plan posting compiled binaries of the renderer?
Regarding the Blender exporter: wouldn't it be a simpler solution to use the exporter of the Mitsuba Blender plugin?
That creates a nice XML file and exports the meshes into PLY files also neatly organised.
In a second step, the XML could be brought in any format you require for Luisa and the materials replaced with your definition.
As I can't try Luisa render for now, I don't know what format you are using but maybe this way would be easier and quicker to implement a translator.

**Reply by Mike-Leo-Smith** [reply]
Thanks for the info about Mitsuba's Blender exporter! Luisa now supports a JSON-based scene format, which might make Mitsuba -> Luisa scene conversion easier, given that the hard part of Blender -> Mitsuba is done.

For binary releases, I can let GitHub Actions produce compiled executables for you to download. There might still be some system-dependent DLLs to handle but let's try first.

**Reply by orarugogerincufelpattano** [reply]
> Thanks for the info about Mitsuba's Blender exporter! Luisa now supports a JSON-based scene format, which might make Mitsuba -> Luisa scene conversion easier, given that the hard part of Blender -> Mitsuba is done.

I'm glad when I could help. I was playing around a bit with Mitsuba and found the exported data very usable and easy to operate with.

> For binary releases, I can let GitHub Actions produce compiled executables for you to download. There might still be some system-dependent DLLs to handle but let's try first.

That would be fantastic! I'm so limited due to the admin restrictions that it's hard to test new ideas.
I would like to test LuisarRender on some light pipe CAD data and see if it can simulate the very challenging light distribution of such a light guide.

**Reply by Mike-Leo-Smith** [reply]
The binaries are here now: https://github.com/LuisaGroup/LuisaRender/actions/runs/7765307417

Just scroll to the bottom of the page and you would see the artifacts. Let me know if any DLLs are missing/unmatched on your system and I'll try to fix them.

**Reply by orarugogerincufelpattano** [reply]
Hey Mike, fantastic! Thank you very much!
I will try to find time on Monday to test it and report back. Have a nice Weekend! BR, Steve

**Reply by orarugogerincufelpattano** [reply]
I downloaded the windows 2022 versions but there is no executable files in them.

Then I took my old Xeon workstation and put my old RTX2070 card in it and installed Windows 10 + VS 2022. 
I cloned the git repository and added all the required packages (Cuda, Vulcan, etc.)
I don't know much about compiling and even less about Visual Studio but at the end I was able to compile everything without errors.
Unfortunately when I try to start the executable files, I get the 0Xc0000142 windows error. (I have to figure out why.)
I also tried to use the "Install" option in Visual Studio to generate a slimmed down version of the compiled software and I get the same problem: all executable files are missing.
Is that some compilation or configuration problem?

**Reply by Mike-Leo-Smith** [reply]
OMG, the mistakes are on my side. I just realized that I didn't correctly handle the CMake install commands after some recent changes. Only the LuisaCompute files are copied during installation but not the LuisaRender executables. Please wait for me to fix that tomorrow. But I'm not sure if this would also fix the `0Xc0000142` error since it might have multiple possible causes.

**Reply by orarugogerincufelpattano** [reply]
Please take your time, no stress. :-) 
Judging by the amount of search results and the lack of a clear answer, it looks like the 0Xc0000142 is a general windows error and might not have anything to do with Luisa at all. I keep testing how to solve it.
(I'm a happy Manjaro user for a long time and it is a pain to work with Windows...)

**Reply by Mike-Leo-Smith** [reply]
Hi @orarugogerincufelpattano 

I have updated the CMake install logic. Now LuisaRender's executables are correctly installed to the same directory as the DLLs. Please go to https://github.com/LuisaGroup/LuisaRender/actions/runs/7818889475 to download the latest compiled binaries (at the bottom of the page) or `git pull` and recompile your local clone.

Thanks for your effort in testing and please let me know if any issues persist :-)

**Reply by orarugogerincufelpattano** [reply]
Many thanks! 
I compiled it and also downloaded your compiled version. Both are missing **tinyexr.dll**.
I checked it on github but I'm not sure if I need to compile it my self or can you provide a compiled version? 
At least the 0Xc0000142 did not popped up yet...

**Reply by Mike-Leo-Smith** [reply]
Oh sorry I forgot to install it. I’ll fix it soon.

**Reply by Mike-Leo-Smith** [reply]
Fixed (again): https://github.com/LuisaGroup/LuisaRender/actions/runs/7830524962

**Reply by orarugogerincufelpattano** [reply]
Thanks!. Now it is complaining about a missing zlib1__.dll
(I only tried your copy. I had no time to clone the repo and compile it myself.)

**Reply by Mike-Leo-Smith** [reply]
Hi @orarugogerincufelpattano

I failed to find which library is linking this DLL so I would appreciate it if you could help provide some more information:

- Which backend did you use?
- Which scene did you try to render?
- When was the missing DLL reported, like when launching the renderer/loading the scene/saving the rendering/etc?

Thanks!

**Reply by orarugogerincufelpattano** [reply]
Hi! I didn't had the time to compile myself so I downloaded the LuisaRender-windows-2022-clang file and extracted it.
(BTW sorry for the noob question, but what is the difference between the cl; clang; clang-cl versions?)
When I try to start **luisa-render-cli.exe** from the command line (dos cmd) without any argument it gives the message that it can't find the zlib1__.dll
Normaly, without arguments it should give a usage screen, or am I wrong?
Maybe tomorrow I will have some time to clone and compile again myself. 
As soon as I have some time, I will download a test scene and try to test it.

**Reply by orarugogerincufelpattano** [reply]
It seems like ZLIB is required by NVidia software like Cuda and also Rust. 
I downloaded a ZLIB library and placed it im my path. That solved the issue.
Now there are no more complaints about missing DLLs but now I have that damned 0xc0000142 error back.
I guess it has something to do with the Windows on my PC. I have to keep investigating...

**Reply by Mike-Leo-Smith** [reply]
Thanks for the information and investigation!

> what is the difference between the cl; clang; clang-cl versions?

They are compiled with different compilers: `cl` for MSVC, `clang` for LLVM Clang with GNU-like command lines, and `clang-cl` for Clang with MSVC-flavored command lines. They should have no functional differences so downloading any should be fine.

> Normally, without arguments it should give a usage screen, or am I wrong?

Yes it should give a usage screen.

> It seems like ZLIB is required by NVidia software like Cuda and also Rust.

It may indicate that the GitHub Actions PCs and yours have mismatched versions of CUDA/system libraries. This could also be the cause of the mysterious 0xc0000142 error (but I'm not sure).

Could you provide the system/graphics driver/cuda versions? Such information would be very helpful for us to track down the issues.

**Reply by orarugogerincufelpattano** [reply]
> Could you provide the system/graphics driver/cuda versions? Such information would be very helpful for us to track down the issues.

Sure, as I don't know much about Windows, so you might need to give me some instructions. But here is what I can figure out alone:
Processor: Intel(R) Xeon(R) CPU W3550  @ 3.07GHz   3.06 GHz
RAM: 14GB
OS: Windows 10 Pro - 22H2
Nvidia RTX 2070 
Studio Driver 551,23 - DCH
Cuda: 12.4.74
DirectX 12.0
For any more, you need to give me some instructions.

**Reply by Mike-Leo-Smith** [reply]
Thanks! The info is quite adequate.

I don't see apparent issues in the versions. Then it may be a privilege issue, like, the system blocked the unregistered DLLs in the binary release. If this is the true cause, the following command might help (run with `cmd.exe`):
```cmd
for %1 in (path\to\LuisaRender\release\*.dll) do regsvr32.exe /s %1
```

Meanwhile, I'll set up a similar environment to see if I can reproduce the issues.

**Reply by Mike-Leo-Smith** [reply]
BTW, just to be sure, do you have the same issues with the [`LuisaRender-windows-2022-cl`](https://github.com/LuisaGroup/LuisaRender/actions/runs/7862650286/artifacts/1236633889) version?

After some searching it turns out that when using Rust with non-MSVC toolchains there could be the `zlib1__.dll` issue (like [this](https://github.com/rust-lang/rust/issues/85422)) and I'm curious if it also could cause the 0xc0000142 problem.

**Comment by orarugogerincufelpattano** [comment]
> BTW, just to be sure, do you have the same issues with the [LuisaRender-windows-2022-cl](https://github.com/LuisaGroup/LuisaRender/actions/runs/7862650286/artifacts/1236633889) version?

Yes, same 0xc0000142 error.
I tried to run regsvr31 on both version (CL & CLANG ) but that didn't helped.
I have a german windows, could that be some local settings that cause this problem?
(That's why I hate windows so much. In comparison to Manjaro or any other Linux or Unix version, it is so badly programmed. It's a shame that it became so widely used...)
The Zlib link you posted was exactly where I landed when I was looking for answers. 
I downloaded zlib123dllx64.zip and placed in the path as described but I had to rename zlibwapi.dll to zlib1__.dll.
Maybe that is causing the problem? Do you have an alternative location where I can download a proper zlib1__.dll?

I'm currently testing on myprivate windows pc. It will be interesting if get it to run here, if I'm able to transfer the installation to my work PC where I have no admin rights.... but one step at a time. =)

**Reply by Mike-Leo-Smith** [reply]
> The Zlib link you posted was exactly where I landed when I was looking for answers.
> I downloaded zlib123dllx64.zip and placed in the path as described but I had to rename zlibwapi.dll to zlib1__.dll.
> Maybe that is causing the problem? Do you have an alternative location where I can download a proper zlib1__.dll?

I set up a Windows environment and reproduced the missing `zlib1__.dll` issue with the `*-clang` version downloaded from GitHub Actions. It was fixed with [this DLL](https://github.com/rust-lang/rust/files/6550164/zlib1__.zip) and then everything worked fine after I put the extracted `zlib1__.dll` under the same directory as `luisa-render-cli.exe`. The `*-cl` version had no issues from the beginning on my side.

> I have a german windows, could that be some local settings that cause this problem?

I'm not sure (I'm not an experienced Windows programmer neither). Are you putting the executables under some non-ASCII paths?

I was running a Windows 11 PC. Some docs say 0xc0000142 would arise due to incompatible Windows versions. So I'll try if I can reproduce it on Windows 10.

**Reply by orarugogerincufelpattano** [reply]
I hate windows so deeply, there is no words for it. It's the biggest garbage of all times...  (Sorry for my rant but I'm mad that this stupid OS is not able to give proper error messages...)
I put your zlib1__.dll in to the cl directory which you compiled. But I still get the same 0xc0000142 error.
I even copied the whole cl directory on another windows 10 notebook (no RTX) and the error remained the same.
Then I changed the local and language settings to Uk and English. Error remains the same.
It drives me mad.
Is there a way to debug in Visual Studio? Unfortunately I don't know anything about VS so I'm a bit lost.

I think, there is nothing wrong with your code, there must be some trivial problem with Windows, that causes this stupid error.

Oh, and no, the directory containing the executable has no special characters in it's name.

**Reply by orarugogerincufelpattano** [reply]
I was reading a bit about debugging in Visual Studio and tried to debug the luisa-render-cli.exe file. 
When I start the debugging, it stops with the following error:
Ausnahmefehler bei 0x00007FFEBA92A708 (lc-ext-eastl.dll) in luisa-render-cli.exe: 0xC000001D: Illegal Instruction
(Exception Error at...) Does this help you in any way?

I had no time to clone the Repo again and compile it myself but as soon as I have time I will do that because then I can link the exe to the source at debugging. That might help to find the problem.

**Reply by orarugogerincufelpattano** [reply]
![image](https://github.com/LuisaGroup/LuisaRender/assets/66918646/b8213789-09e2-4fc2-90b5-5ca3a1835643)

I cloned the repository and recompiled it. 
This is the code segment where the error happens. 
I don't know anything about C++ so I can't say why it's causing a problem... maybe you understand it. :-)

**Reply by Mike-Leo-Smith** [reply]
Thanks a lot!

This is an error inside the third-party `mimalloc` library. I have no idea why this simple bit operation is causing an illegal instructions error, at my first glance. But knowing where it’s happening really helps a lot. I’ll keep investigating it.

**Reply by Mike-Leo-Smith** [reply]
One possibility is that, the compiler is using some new instruction sets that your CPU doesn’t support. I will try to identify these instructions.

**Reply by orarugogerincufelpattano** [reply]
That's absolutely possible as it's an old Xeon workstation. The laptop I tested is a newer model (2019) but it only has Pentium 5405SU Gold processor (64 bit). 
 
EDIT: I think you had the right idea. I was searching a bit and if I understand it correctly, some bit shift operations are requiring AVX2 instruction set support from the CPU. Neither the W3550 Xeon nor the 5405U Pentium has support for this instruction set. 
That would explain why it crashes. 
In a C++ visual studio project there is a setting to turn the AVX instruction sets off but I was not able to find how to do the same with a GitHub project in visual studio. Maybe you have an idea?

Here is a discussion about this feature, which in VS 2022 obviously has some problems:
[Visual studio instruction sets settings](https://developercommunity.visualstudio.com/t/Cannot-disable-AVX-and-AVX2-in-VS-2022/10497078?space=21&scope=follow&sort=newest)

**Reply by Mike-Leo-Smith** [reply]
Thanks for the info!

I suspect the `avx2` instruction set too. So I am modifying `CMakeLists.txt` in both LuisaCompute and LuisaRender to disable it. This might take me a day or two and then we'll see if it would solve the issues :-)

**Reply by orarugogerincufelpattano** [reply]
Cool, thanks and sorry that I bring such issues just because I want to test the renderer on old hardware.

**Reply by Mike-Leo-Smith** [reply]
No worries. It's also a good chance for us to improve compatibility.

**Reply by Mike-Leo-Smith** [reply]
Hi @orarugogerincufelpattano 

Here is the new version without `avx2` enabled: https://github.com/LuisaGroup/LuisaRender/actions/runs/8008910888/artifacts/1267620431

**Reply by Mike-Leo-Smith** [reply]
If you would like to compile from source, you would need to also update the submodules:
```bash
git pull && git submodule update --init --recursive
```

**Reply by orarugogerincufelpattano** [reply]
Thanks! I compiled it and if I start it, I get the usage screen! Fantastic!
I downloaded a sample scene and placed it in to the bin folder and tried to render. That gave me a debbuger error.
Is there a documentation on how to render with the renderer? For example what backends are available and so on?

EDIT: I used "-b cuda" as backend and the rendering started and finished. =)
There was an error message at the end, but the image was still done.
 [warning] Failed to allocate 34560000 bytes from CUDAHostBufferPool. Falling back to ad-hoc allocation. [D:\a\LuisaRender\LuisaRender\src\compute\src\backends\cuda\cuda_host_buffer_pool.cpp:35]
(I don't have a D: drive in my machine so I don't really understand the message...)

I'm really looking forward to read more abut the renderer and materials and so on.

**Reply by Mike-Leo-Smith** [reply]
Very happy to hear the good news!

> For example what backends are available and so on?

If you are using an NVIDIA card on Windows, then the `cuda`, `dx`, and `cpu` backends are available.

> EDIT: I used "-b cuda" as backend and the rendering started and finished. =)
> There was an error message at the end, but the image was still done.
> [warning] Failed to allocate 34560000 bytes from CUDAHostBufferPool. Falling back to ad-hoc allocation. [D:\a\LuisaRender\LuisaRender\src\compute\src\backends\cuda\cuda_host_buffer_pool.cpp:35]
> (I don't have a D: drive in my machine so I don't really understand the message...)

The warning is just fine (as explained [here](https://github.com/LuisaGroup/LuisaRender/issues/10), it just indicates some minor internal optimization failure). The `D:\...` path is the source file path on the GitHub Actions machiine where the binary release was built.

> I'm really looking forward to read more abut the renderer and materials and so on.

Sorry we don't have a comprehensive doc/tutorial at present. The usage of renderer itself is quite simple, with all available options printed in the help screen. The underlying [compute](https://github.com/LuisaGroup/LuisaCompute) library, and its design and implementation mechanism, are described in our [paper](https://dl.acm.org/doi/10.1145/3550454.3555463), which might be a good reference if you're interested about how the high-level rendering process is materialized into low-level compute shaders and API calls. But still sorry that we have not prepared a good doc currently.

**Reply by orarugogerincufelpattano** [reply]
Thanks and no problem. I'm studying the sample scenes and analyzing the Tungsten converter.
I have some scene specific questions, can I ask them or am I nerving already too much?

**Reply by Mike-Leo-Smith** [reply]
Please feel free to ask anything you'd like to know :-)

**Reply by orarugogerincufelpattano** [reply]
Thanks! I really appreciate it! I'll open up a new discussion as this was more about the installation.
